### DISCUSSION TITLE
How to implement fire-and-forget pattern

### DISCUSSION THREAD

**Discussion by oskin1** [libp2p/rust-libp2p#3786] [Q&A] [answered]
Some sub-protocols in my application require sending to concrete peers just a single message. Does libp2p-rust allow to send a message to some `PeerId` without negotiating a substream?

**Comment by thomaseizinger** [comment, marked as answer]
If you use the `Upgrade::V1Lazy` strategy, negotiating substreams is 0RTT. The `OneShotHandler` may be useful for your usecase: https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/handler/one_shot.rs

**Reply by oskin1** [reply]
@thomaseizinger Thanks!
Where do I specify upgrade strategy? Also, can I somehow combine `Upgrade::V1Lazy` and `Upgrade::V1` to support both one-shot protocols and protocols that require negotiation?

**Reply by thomaseizinger** [reply]
`V1Lazy` is backwards-compatible to `V1`, see its docs: https://docs.rs/libp2p/0.51.3/libp2p/core/upgrade/enum.Version.html

See the `identify` example for how to use it: https://github.com/libp2p/rust-libp2p/blob/master/examples/identify/src/main.rs#L55

**Reply by oskin1** [reply]
@thomaseizinger Yes, in the documentation for `V1Lazy` it is said that "This strategy is only applicable for the node with the role of “dialer” in the negotiation and only if the dialer supports just a single application protocol.", while in my case the node should still support both dialing and listening for a number of other protocols.

**Reply by thomaseizinger** [reply]
The configuration option is ignored for listening so you are all good. You can set it to `V1Lazy` without problems. In fact, we want to make it the default: https://github.com/libp2p/rust-libp2p/issues/3563
